python/src/lib.rs

```diff
@@ -232,6 +232,116 @@ impl RawDeltaTable {
             .block_on(self._table.update_incremental())
             .map_err(PyDeltaTableError::from_raw)
     }
+
+    pub fn dataset_partitions<'py>(
+        &mut self,
+        py: Python<'py>,
+        partition_filters: Option<Vec<(&str, &str, PartitionFilterValue)>>,
+    ) -> PyResult<Vec<(String, Option<&'py PyAny>)>> {
+        let path_set = match partition_filters {
+            Some(filters) => Some(HashSet::<_>::from_iter(
+                self.files_by_partitions(filters)?.iter().cloned(),
+            )),
+            None => None,
+        };
+
+        self._table
+            .get_file_uris()
+            .zip(self._table.get_partition_values())
+            .zip(self._table.get_stats())
+            .filter(|((path, _), _)| match &path_set {
+                Some(path_set) => path_set.contains(path),
+                None => true,
+            })
+            .map(|((path, partition_values), stats)| {
+                let stats = stats.map_err(PyDeltaTableError::from_raw)?;
+                let expression = filestats_to_expression(py, partition_values, stats)?;
+                Ok((path, expression))
+            })
+            .collect()
+    }
+}
+
+fn json_value_to_py(value: &serde_json::Value, py: Python) -> PyObject {
```

Comment:
I'm struggling to understand why this function needs to handle only some of the branches of the `serde_json::Value` enum rather than everything or nothing. As far as I can tell `ColumnValueStat` in this case is always going to be `Value(DeltaDataTypeLong)` (i.e. `i64`).

I don't really have feedback here, just trying to understand

Reply:
`ColumnCountStat::Value` (which is used for null counts) is always going to be `Value(DeltaDataTypeLong)`. But `ColumnValueStat::Value` could be any JSON primitive value, depending on the type of column. Minimum values of int columns will be ints, strings will be strings. See an example here:

https://github.com/delta-io/delta-rs/blob/57133e161e03a34d8902abd210dca80731ce00d1/rust/tests/data/COVID-19_NYT/_delta_log/00000000000000000000.json#L4

Which turns into the following partition expression:

```python
<pyarrow.dataset.Expression ((((((((((((((((date >= "2020-01-21") and (fips >= 1001)) and (state >= "Alabama")) and (county >= "Abbeville")) and (deaths >= 0)) and (cases >= 0)) and (state <= "Wyoming")) and (date <= "2020-05-19")) and (county <= "Ziebach")) and (fips <= 78030)) and (cases <= 198114)) and (deaths <= 20298)) and is_valid(date)) and is_valid(state)) and is_valid(county)) and is_valid(cases))>
```
AFAIK I shouldn't need to handle array or object JSON values here. Maybe I should panic on them instead?

Reply:
I think it would be good to throw an exception in case of unexpected json value types.